File: apps/api/models/water.py
"""
models/water.py — ledger eau et screening hydrique (PR-08).

Tranche A : imports/activités/permis/zones de stress (migration 036).
Tranche B : screenings versionnés, cibles et actions (migration 037).

Champs snake_case alignés sur les colonnes SQL. Invariants portés par les
types (pas seulement par la documentation) :

- `WaterScreeningData.risk_category` et `confidence` sont DEUX champs
  distincts — aucun champ combiné n'existe (précédent CRMA, contrats §6
  Wave 4) : l'UI ne peut pas afficher un « risque net » par accident.
- `method_code` nomme TOUJOURS la méthode géométrique réellement utilisée
  (geojson_point_in_polygon_v1) — jamais présentée comme ST_Intersects.
- `iro_signal` n'est jamais une décision : le modèle ne porte aucun champ de
  « matérialité » — seulement le signal, sa justification et son auteur.
- Le référentiel de zones expose sa licence (`display_allowed`,
  `derived_use_allowed`, `attribution_text`) lue via license_policy.evaluate —
  jamais dénormalisée en base.
"""

from __future__ import annotations

from datetime import date, datetime
from typing import Any, Literal

from pydantic import BaseModel, Field

WaterActivityType = Literal["withdrawal", "consumption", "discharge"]
WaterSourceType = Literal["surface", "groundwater", "municipal", "seawater", "other"]
WaterDataStatus = Literal["verified", "estimated", "manual", "inferred"]
ReviewStatus = Literal["pending", "accepted", "flagged"]
ImportStatus = Literal["pending", "validated", "rejected"]
PermitType = Literal["withdrawal", "discharge", "operation", "other"]
PermitStatus = Literal["active", "expired", "revoked"]
AreaKind = Literal["basin", "aquifer", "administrative", "custom"]
StressCategory = Literal["low", "low_medium", "medium_high", "high", "extremely_high"]
TargetType = Literal[
    "withdrawal_reduction", "consumption_reduction", "discharge_quality", "reuse_rate", "other"
]
TargetStatus = Literal["draft", "active", "achieved", "abandoned"]
WaterActionType = Literal[
    "efficiency", "reuse", "recycling", "sourcing_change", "treatment", "monitoring", "other"
]
WaterActionStatus = Literal["planned", "in_progress", "completed", "cancelled"]


# ---------------------------------------------------------------------------
# Imports (idempotence de contenu) et activités
# ---------------------------------------------------------------------------

class WaterImportRequest(BaseModel):
    """Import CSV d'activités eau. `csv_text` = contenu brut (colonnes :
    site_id, activity_type, source_type, quantity_m3, period_start,
    period_end). Idempotent par CONTENU (sha256 — patron purchase_imports) ET
    par ligne (clé naturelle UNIQUE en base)."""

    filename: str = Field(min_length=1, max_length=300)
    csv_text: str


class WaterImportResponse(BaseModel):
    id: int
    company_id: int
    filename: str
    sha256: str
    row_count: int
    accepted_count: int
    rejected_count: int
    status: ImportStatus
    imported_by: int | None
    imported_at: datetime
    already_imported: bool = False
    errors: list[str] = Field(default_factory=list)


class WaterImportReviewRequest(BaseModel):
    accept: bool


class WaterActivityCreate(BaseModel):
    site_id: int
    activity_type: WaterActivityType
    source_type: WaterSourceType = "other"
    quantity_m3: float = Field(ge=0)
    period_start: date
    period_end: date
    data_status: WaterDataStatus = "manual"
    evidence_artifact_id: int | None = None


class WaterActivityResponse(BaseModel):
    id: int
    company_id: int
    site_id: int
    activity_type: WaterActivityType
    source_type: WaterSourceType
    quantity_m3: float
    period_start: date
    period_end: date
    import_id: int | None
    data_status: WaterDataStatus
    evidence_artifact_id: int | None
    review_status: ReviewStatus
    created_at: datetime
    updated_at: datetime


class WaterActivityListResponse(BaseModel):
    items: list[WaterActivityResponse]
    total: int
    limit: int
    offset: int


class WaterActivityReviewRequest(BaseModel):
    accept: bool


# ---------------------------------------------------------------------------
# Permis
# ---------------------------------------------------------------------------

class WaterPermitCreate(BaseModel):
    site_id: int
    permit_type: PermitType
    permit_reference: str | None = Field(default=None, max_length=200)
    authorized_volume_m3: float | None = Field(default=None, ge=0)
    valid_from: date | None = None
    valid_to: date | None = None
    issuing_authority: str | None = Field(default=None, max_length=300)
    evidence_artifact_id: int | None = None
    status: PermitStatus = "active"
    notes: str | None = Field(default=None, max_length=2000)


class WaterPermitResponse(BaseModel):
    id: int
    company_id: int
    site_id: int
    permit_type: PermitType
    permit_reference: str | None
    authorized_volume_m3: float | None
    valid_from: date | None
    valid_to: date | None
    issuing_authority: str | None
    evidence_artifact_id: int | None
    status: PermitStatus
    review_status: ReviewStatus
    notes: str | None
    created_at: datetime
    updated_at: datetime


class WaterPermitListResponse(BaseModel):
    items: list[WaterPermitResponse]
    total: int
    limit: int
    offset: int


# ---------------------------------------------------------------------------
# Référentiel de zones de stress hydrique (toujours sourcé, licence surfacée)
# ---------------------------------------------------------------------------

class WaterRiskAreaResponse(BaseModel):
    """Zone de stress hydrique. Quand la licence de la source refuse
    l'affichage, `baseline_stress_category` est retirée CÔTÉ SERVEUR
    (`value_withheld=True`) — la valeur ne quitte jamais le backend (précédent
    market_observations PR-07)."""

    id: int
    company_id: int | None
    code: str
    label: str
    area_kind: AreaKind
    scenario_code: str
    horizon_year: int | None
    baseline_stress_category: StressCategory | None
    bbox_min_lat: float
    bbox_max_lat: float
    bbox_min_lon: float
    bbox_max_lon: float
    source_release_id: int
    evidence_artifact_id: int | None
    data_status: WaterDataStatus
    source_code: str | None = None
    # Licence lue à l'usage via license_policy.evaluate (jamais stockée) :
    display_allowed: bool = True
    derived_use_allowed: bool = True
    license_reasons: list[str] = Field(default_factory=list)
    attribution_text: str | None = None
    value_withheld: bool = False
    created_at: datetime
    updated_at: datetime


class WaterRiskAreaListResponse(BaseModel):
    items: list[WaterRiskAreaResponse]
    total: int
    limit: int
    offset: int


# ---------------------------------------------------------------------------
# Screening (tranche B) — risque et confiance SÉPARÉS, méthode explicite
# ---------------------------------------------------------------------------

class WaterScreeningRequest(BaseModel):
    site_id: int
    scenario_code: str = Field(default="baseline", min_length=1, max_length=100)


class MatchedAreaResult(BaseModel):
    """Trace d'appariement d'UNE zone — inspectable, jamais un total opaque."""

    area_id: int
    code: str
    label: str
    area_kind: AreaKind
    stress_category: StressCategory
    data_status: WaterDataStatus
    bbox_candidate: bool
    matched: bool
    method_code: str
    prefilter_code: str


class WaterScreeningData(BaseModel):
    """Résultat métier d'un screening — `data` de l'enveloppe analytique.

    `risk_category` (intensité, issue des zones appariées) et `confidence`
    (solidité du socle documentaire, portée par meta.quality) ne se combinent
    JAMAIS. `risk_category=None` signifie « aucune zone connue appariée », ce
    qui n'est PAS un risque nul — les warnings le disent explicitement."""

    screening_id: int
    site_id: int
    scenario_code: str
    method_code: str
    methodology_code: str
    methodology_version: str
    risk_category: StressCategory | None
    matched_areas: list[MatchedAreaResult]
    candidate_area_count: int
    matched_area_count: int
    iro_signal: bool
    iro_signal_rationale: str | None
    input_fingerprint: str
    calculated_at: datetime | None


class WaterScreeningSummary(BaseModel):
    id: int
    company_id: int
    site_id: int
    methodology_code: str
    methodology_version: str
    method_code: str
    scenario_code: str
    risk_category: StressCategory | None
    confidence: float | None
    coverage_pct: float | None
    warnings: list[Any] = Field(default_factory=list)
    iro_signal: bool
    iro_signal_rationale: str | None
    iro_signal_by: int | None
    iro_signal_at: datetime | None
    input_fingerprint: str
    calculated_at: datetime
    calculated_by: int | None


class WaterScreeningListResponse(BaseModel):
    items: list[WaterScreeningSummary]
    total: int
    limit: int
    offset: int


class IroSignalRequest(BaseModel):
    """Pose le signal « à examiner comme IRO » sur un screening. Un GESTE
    humain avec justification obligatoire — jamais une décision de matérialité
    (la promotion effective en IRO est l'affaire de PR-10)."""

    rationale: str = Field(min_length=1, max_length=2000)


# ---------------------------------------------------------------------------
# Cibles et actions
# ---------------------------------------------------------------------------

class WaterTargetCreate(BaseModel):
    target_type: TargetType
    title: str = Field(min_length=1, max_length=300)
    site_id: int | None = None
    screening_id: int | None = None
    description: str | None = Field(default=None, max_length=2000)
    baseline_year: int | None = Field(default=None, ge=1900, le=2200)
    target_year: int | None = Field(default=None, ge=1900, le=2200)
    baseline_value_m3: float | None = Field(default=None, ge=0)
    target_value_m3: float | None = Field(default=None, ge=0)
    status: TargetStatus = "draft"
    notes: str | None = Field(default=None, max_length=2000)


class WaterTargetResponse(BaseModel):
    id: int
    company_id: int
    site_id: int | None
    screening_id: int | None
    target_type: TargetType
    title: str
    description: str | None
    baseline_year: int | None
    target_year: int | None
    baseline_value_m3: float | None
    target_value_m3: float | None
    status: TargetStatus
    review_status: ReviewStatus
    notes: str | None
    created_at: datetime
    updated_at: datetime


class WaterTargetListResponse(BaseModel):
    items: list[WaterTargetResponse]
    total: int
    limit: int
    offset: int


class WaterActionCreate(BaseModel):
    action_type: WaterActionType
    title: str = Field(min_length=1, max_length=300)
    site_id: int | None = None
    screening_id: int | None = None
    target_id: int | None = None
    description: str | None = Field(default=None, max_length=2000)
    status: WaterActionStatus = "planned"
    owner: str | None = Field(default=None, max_length=200)
    due_date: date | None = None
    expected_effect: str | None = Field(default=None, max_length=2000)
    expected_reduction_m3: float | None = Field(default=None, ge=0)


class WaterActionResponse(BaseModel):
    id: int
    company_id: int
    site_id: int | None
    screening_id: int | None
    target_id: int | None
    action_type: WaterActionType
    title: str
    description: str | None
    status: WaterActionStatus
    owner: str | None
    due_date: date | None
    completed_at: datetime | None
    expected_effect: str | None
    expected_reduction_m3: float | None
    review_status: ReviewStatus
    created_at: datetime
    updated_at: datetime


class WaterActionListResponse(BaseModel):
    items: list[WaterActionResponse]
    total: int
    limit: int
    offset: int

File: apps/carbon/lib/api/water.ts
/**
 * lib/api/water.ts — client des endpoints `/water/*` et `/sites/geo*`
 * (PR-08 : géospatial des sites, ledger eau, screening hydrique).
 *
 * Réutilise `API_BASE_URL` + le porteur de token en mémoire de `lib/api.ts`.
 * Types en snake_case, miroir de `apps/api/models/geo.py` / `models/water.py`.
 *
 * Invariants du domaine lisibles dans ces types :
 *
 * 1. **Gate de géocodage.** `SiteGeo.position_usable` n'est vrai qu'après
 *    acceptation HUMAINE — l'UI n'a aucun moyen d'afficher une position
 *    « utilisable » qui ne l'est pas.
 * 2. **Méthode géométrique nommée.** `method_code` du screening est
 *    `geojson_point_in_polygon_v1` (bbox = pré-filtre) — jamais présentée
 *    comme PostGIS/ST_Intersects, et l'UI l'affiche telle quelle.
 * 3. **Risque ≠ confiance.** `risk_category` (data) et `confidence`
 *    (meta.quality) ne partagent aucun champ combiné.
 * 4. **Schéma pas encore migré.** Le backend répond 503 `schema_not_ready`
 *    tant que 036/037 ne sont pas appliquées en production : le client le
 *    traduit en `SchemaNotReadyError`, que chaque page BETA rend comme
 *    « initialisation du schéma en cours » — jamais une erreur brute.
 */

import { API_BASE_URL, getAuthToken } from "@/lib/api";

// ---------------------------------------------------------------------------
// Types (miroir models/geo.py + models/water.py)
// ---------------------------------------------------------------------------

export type ReviewStatus = "pending" | "accepted" | "flagged";
export type CandidateStatus = "proposed" | "accepted" | "rejected";
export type GeocodePrecision = "exact" | "street" | "city" | "country" | "manual";
export type WaterActivityType = "withdrawal" | "consumption" | "discharge";
export type StressCategory = "low" | "low_medium" | "medium_high" | "high" | "extremely_high";

export interface SiteGeo {
  id: number;
  company_id: number;
  name: string;
  location: string | null;
  latitude: number | null;
  longitude: number | null;
  geocode_precision: GeocodePrecision | null;
  geocode_provider: string | null;
  geocode_review_status: ReviewStatus;
  position_usable: boolean;
}

export interface GeocodeCandidate {
  id: number;
  site_id: number;
  provider: string;
  provider_ref: string | null;
  latitude: number;
  longitude: number;
  precision: GeocodePrecision | null;
  method_code: string;
  status: CandidateStatus;
  review_note: string | null;
  reviewed_by: number | null;
  reviewed_at: string | null;
  created_at: string;
}

export interface WaterActivity {
  id: number;
  site_id: number;
  activity_type: WaterActivityType;
  source_type: string;
  quantity_m3: number;
  period_start: string;
  period_end: string;
  data_status: string;
  review_status: ReviewStatus;
  evidence_artifact_id: number | null;
}

export interface WaterImportResult {
  id: number;
  filename: string;
  row_count: number;
  accepted_count: number;
  rejected_count: number;
  status: "pending" | "validated" | "rejected";
  already_imported: boolean;
  errors: string[];
}

export interface WaterPermit {
  id: number;
  site_id: number;
  permit_type: string;
  permit_reference: string | null;
  authorized_volume_m3: number | null;
  valid_from: string | null;
  valid_to: string | null;
  issuing_authority: string | null;
  evidence_artifact_id: number | null;
  status: "active" | "expired" | "revoked";
  review_status: ReviewStatus;
}

export interface WaterRiskArea {
  id: number;
  company_id: number | null;
  code: string;
  label: string;
  area_kind: string;
  scenario_code: string;
  /** `null` quand la licence refuse l'affichage (`value_withheld`). */
  baseline_stress_category: StressCategory | null;
  source_code: string | null;
  display_allowed: boolean;
  derived_use_allowed: boolean;
  license_reasons: string[];
  attribution_text: string | null;
  value_withheld: boolean;
  data_status: string;
}

export interface MatchedArea {
  area_id: number;
  code: string;
  label: string;
  stress_category: StressCategory;
  data_status: string;
  bbox_candidate: boolean;
  matched: boolean;
  method_code: string;
  prefilter_code: string;
}

export interface WaterScreeningData {
  screening_id: number;
  site_id: number;
  scenario_code: string;
  method_code: string;
  methodology_code: string;
  methodology_version: string;
  risk_category: StressCategory | null;
  matched_areas: MatchedArea[];
  candidate_area_count: number;
  matched_area_count: number;
  iro_signal: boolean;
  input_fingerprint: string;
  calculated_at: string | null;
}

export interface WaterScreeningEnvelope {
  data: WaterScreeningData;
  meta: {
    as_of: string | null;
    status: string;
    method: { code: string; version: string };
    quality: { confidence: number | null; coverage_pct: number | null; warnings: string[] };
  };
  evidence: { source_code: string | null; note: string | null }[];
}

export interface WaterScreeningSummary {
  id: number;
  site_id: number;
  method_code: string;
  methodology_code: string;
  methodology_version: string;
  scenario_code: string;
  risk_category: StressCategory | null;
  confidence: number | null;
  coverage_pct: number | null;
  warnings: string[];
  iro_signal: boolean;
  iro_signal_rationale: string | null;
  calculated_at: string;
}

export interface WaterTarget {
  id: number;
  site_id: number | null;
  screening_id: number | null;
  target_type: string;
  title: string;
  baseline_year: number | null;
  target_year: number | null;
  baseline_value_m3: number | null;
  target_value_m3: number | null;
  status: string;
  review_status: ReviewStatus;
}

export interface WaterAction {
  id: number;
  site_id: number | null;
  screening_id: number | null;
  target_id: number | null;
  action_type: string;
  title: string;
  status: string;
  expected_reduction_m3: number | null;
  review_status: ReviewStatus;
}

export interface Paginated<T> {
  items: T[];
  total: number;
  limit: number;
  offset: number;
}

// ---------------------------------------------------------------------------
// Client — avec détection contractuelle de `schema_not_ready`
// ---------------------------------------------------------------------------

/** 503 `schema_not_ready` : les migrations 036/037 ne sont pas encore
 * appliquées. Les pages BETA rendent « initialisation du schéma en cours ». */
export class SchemaNotReadyError extends Error {
  constructor() {
    super("schema_not_ready");
    this.name = "SchemaNotReadyError";
  }
}

function authHeaders(): Record<string, string> {
  const token = getAuthToken();
  return token ? { Authorization: `Bearer ${token}` } : {};
}

async function handle<T>(res: Response, path: string): Promise<T> {
  if (res.status === 503) {
    let detail: unknown = null;
    try {
      detail = ((await res.json()) as { detail?: unknown }).detail;
    } catch {
      /* corps non-JSON : erreur générique ci-dessous */
    }
    if (detail === "schema_not_ready") throw new SchemaNotReadyError();
  }
  if (!res.ok) {
    let detail = "";
    try {
      detail = String(((await res.json()) as { detail?: unknown }).detail ?? "");
    } catch {
      /* garder le message générique */
    }
    throw new Error(detail || `API ${res.status} on ${path}`);
  }
  return (await res.json()) as T;
}

async function get<T>(path: string, signal?: AbortSignal): Promise<T> {
  const res = await fetch(`${API_BASE_URL}${path}`, {
    method: "GET",
    headers: { Accept: "application/json", ...authHeaders() },
    credentials: "include",
    signal,
  });
  return handle<T>(res, path);
}

async function post<T>(path: string, body: unknown): Promise<T> {
  const res = await fetch(`${API_BASE_URL}${path}`, {
    method: "POST",
    headers: {
      Accept: "application/json",
      "Content-Type": "application/json",
      ...authHeaders(),
    },
    credentials: "include",
    body: JSON.stringify(body),
  });
  return handle<T>(res, path);
}

// ── Sites géo & candidats ────────────────────────────────────────────────────

export function fetchSitesGeo(signal?: AbortSignal): Promise<Paginated<SiteGeo>> {
  return get<Paginated<SiteGeo>>("/sites/geo?limit=200", signal);
}

export function fetchGeocodeCandidates(
  siteId: number,
  signal?: AbortSignal,
): Promise<Paginated<GeocodeCandidate>> {
  return get<Paginated<GeocodeCandidate>>(`/sites/${siteId}/geocode-candidates?limit=100`, signal);
}

export function proposeGeocodeCandidate(
  siteId: number,
  body: { latitude: number; longitude: number; precision?: GeocodePrecision; provider?: string },
): Promise<GeocodeCandidate> {
  return post<GeocodeCandidate>(`/sites/${siteId}/geocode-candidates`, body);
}

export function reviewGeocodeCandidate(
  siteId: number,
  candidateId: number,
  accept: boolean,
  note?: string,
): Promise<GeocodeCandidate> {
  return post<GeocodeCandidate>(
    `/sites/${siteId}/geocode-candidates/${candidateId}/review`,
    { accept, note: note ?? null },
  );
}

// ── Eau : activités, permis, zones ──────────────────────────────────────────

export function fetchWaterActivities(signal?: AbortSignal): Promise<Paginated<WaterActivity>> {
  return get<Paginated<WaterActivity>>("/water/activities?limit=200", signal);
}

export function importWaterCsv(filename: string, csvText: string): Promise<WaterImportResult> {
  return post<WaterImportResult>("/water/activities/import", {
    filename,
    csv_text: csvText,
  });
}

export function reviewWaterActivity(activityId: number, accept: boolean): Promise<WaterActivity> {
  return post<WaterActivity>(`/water/activities/${activityId}/review`, { accept });
}

export function fetchWaterPermits(signal?: AbortSignal): Promise<Paginated<WaterPermit>> {
  return get<Paginated<WaterPermit>>("/water/permits?limit=200", signal);
}

export function fetchWaterRiskAreas(signal?: AbortSignal): Promise<Paginated<WaterRiskArea>> {
  return get<Paginated<WaterRiskArea>>("/water/risk-areas?limit=200", signal);
}

// ── Screening, cibles, actions ──────────────────────────────────────────────

export function fetchWaterScreenings(
  signal?: AbortSignal,
): Promise<Paginated<WaterScreeningSummary>> {
  return get<Paginated<WaterScreeningSummary>>("/water/screenings?limit=100", signal);
}

export function calculateWaterScreening(
  siteId: number,
  scenarioCode = "baseline",
): Promise<WaterScreeningEnvelope> {
  return post<WaterScreeningEnvelope>("/water/screenings/calculate", {
    site_id: siteId,
    scenario_code: scenarioCode,
  });
}

export function flagScreeningForIro(
  screeningId: number,
  rationale: string,
): Promise<WaterScreeningSummary> {
  return post<WaterScreeningSummary>(`/water/screenings/${screeningId}/flag-for-iro`, {
    rationale,
  });
}

export function fetchWaterTargets(signal?: AbortSignal): Promise<Paginated<WaterTarget>> {
  return get<Paginated<WaterTarget>>("/water/targets?limit=200", signal);
}

export function fetchWaterActions(signal?: AbortSignal): Promise<Paginated<WaterAction>> {
  return get<Paginated<WaterAction>>("/water/actions?limit=200", signal);
}

// ---------------------------------------------------------------------------
// Présentation
// ---------------------------------------------------------------------------

export const STRESS_LABEL: Record<StressCategory, string> = {
  low: "Faible",
  low_medium: "Faible à moyen",
  medium_high: "Moyen à élevé",
  high: "Élevé",
  extremely_high: "Extrême",
};

export const STRESS_TONE: Record<StressCategory, string> = {
  low: "text-emerald-600 dark:text-emerald-400",
  low_medium: "text-lime-600 dark:text-lime-400",
  medium_high: "text-amber-600 dark:text-amber-400",
  high: "text-orange-600 dark:text-orange-400",
  extremely_high: "text-red-600 dark:text-red-400",
};

export const REVIEW_LABEL: Record<ReviewStatus, string> = {
  pending: "En attente de revue",
  accepted: "Acceptée",
  flagged: "Signalée",
};

export const ACTIVITY_LABEL: Record<WaterActivityType, string> = {
  withdrawal: "Prélèvement",
  consumption: "Consommation",
  discharge: "Rejet",
};

/** Formate un volume m³ lisible, ou « n. d. » si absent. */
export function formatM3(value: number | null | undefined): string {
  if (value === null || value === undefined || Number.isNaN(value)) return "n. d.";
  return `${new Intl.NumberFormat("fr-FR", { maximumFractionDigits: 1 }).format(value)} m³`;
}
